### PR TITLE
[SYCL-MLIR]: Add Inliner Pass

### PR DESCRIPTION
**Pull request by etiotto**
This PR extends https://github.com/intel/llvm/pull/7481 and introduces an inliner pass with simple heuristics. 
The inliner pass allows inlining:
  - callees that are marked with the 'alwaysinline' or 'inlinehint' attribute
  - any call edge that would render the callee dead after it is inlined
 
Inline options can be supplied to the pass to select:
   - the inline mode (one of alwaysinline, simple)
   - whether unreachable callees (after inlining) should be removed
   
The inliner collects the number of call edges that were removed (under a statistic).

### CODE REVIEW

File: polygeist/tools/cgeist/driver.cc [hunk 1]
```diff
@@ -342,11 +342,11 @@ static int optimize(mlir::MLIRContext &context,
...
     optPM.addPass(mlir::createCanonicalizerPass(canonicalizerConfig, {}, {}));
     optPM.addPass(mlir::createCSEPass());
-    // Affine must be lowered to enable inlining
+    // Note: affine dialects must be lowered to allow callees containing affine
+    // operations to be inlined.
     if (RaiseToAffine)
       optPM.addPass(mlir::createLowerAffinePass());
-    optPM.addPass(mlir::createCanonicalizerPass(canonicalizerConfig, {}, {}));
-    pm.addPass(sycl::createAlwaysInlinePass());
+    pm.addPass(sycl::createInlinePass());
```
Comment (etiotto) [inline polygeist/tools/cgeist/driver.cc:349]: Should prob. pass the inliner options here.

File: mlir-sycl/lib/Transforms/Inliner.cpp [hunk 2]
```diff
@@ -283,16 +565,37 @@ void Inliner::inlineCallsInSCC(Inliner &Inliner, CallGraphSCC &SCC) {
 
     LLVM_DEBUG(llvm::dbgs() << "** Inline succeeded\n");
```
Comment (sommerlukas) [inline mlir-sycl/lib/Transforms/Inliner.cpp:566]: Would it make sense to have a [pass statistic](https://mlir.llvm.org/docs/PassManagement/#pass-statistics) for the number of successfully inlined calls?

**Review by sommerlukas — CHANGES_REQUESTED** [review]
Just a few small comments.

File: mlir-sycl/lib/Transforms/Inliner.cpp (hunk 2 above)
Comment (etiotto) [reply mlir-sycl/lib/Transforms/Inliner.cpp:569]: Added a statistic to count the number of calls inlined. It is turned on via the `-mlir-pass-statistics` flag (like other MLIR statistics).

**Review by victor-eds — COMMENTED** [review]
We are currently assuming this pass reaches a fixpoint after a single iteration. Could we need more than one? (can address on other PR).

File: mlir-sycl/lib/Transforms/Inliner.cpp [hunk 3]
```diff
@@ -335,32 +640,110 @@ bool AlwaysInliner::shouldInline(ResolvedCall &ResolvedCall) const {
...
+  std::unique_ptr<InlinerBase> Inliner = nullptr;
+  switch (InlineMode) {
+  case sycl::InlineMode::AlwaysInline:
+    Inliner = std::make_unique<class AlwaysInliner>(Ctx, CG, SymTable);
+    break;
+  case sycl::InlineMode::Simple:
+    Inliner = std::make_unique<class Inliner>(Ctx, CG, SymTable);
+    break;
+  case sycl::InlineMode::Aggressive:
+    assert(false && "TODO");
```
[the remark this answers is not in the document]
Comment (etiotto) [reply mlir-sycl/lib/Transforms/Inliner.cpp:711]: done

File: mlir-sycl/lib/Transforms/Inliner.cpp [hunk 4]
```diff
@@ -335,32 +640,110 @@ bool AlwaysInliner::shouldInline(ResolvedCall &ResolvedCall) const {
...
+    return Uses->hasOneUse(ResolvedCall.TgtNode);
+
+  // TODO: extend simple heuristics
+  // - Inline a callee if inlining it would make it dead.
+  //    - consider case where there are more than one call hedge in the same
+  //    caller ?
+  //  - what about inlining in loops (Should allow inlining in inner loop)
+  //  - pass flags to make inliner 'aggressive'?
+  //     + is inliner is aggressive then inline in inner loop as long as
+  //     callee is not too big
```
[the remark this answers is not in the document]
Comment (etiotto) [reply mlir-sycl/lib/Transforms/Inliner.cpp:684]: Actually, this is a leftover comment, removed.

File: mlir-sycl/lib/Transforms/Inliner.cpp [hunk 5]
```diff
@@ -172,47 +281,219 @@ class Inliner : public InlinerInterface {
...
+  for (CallGraphNode *CGN : Uses.TopLevelUses)
+    --DiscardableSymNodeUses[CGN];
+  for (auto &It : Uses.InnerUses)
+    DiscardableSymNodeUses[It.first] -= It.second;
+
+  for (CallGraphNode *CGN : Uses.TopLevelUses) {
+    Operation *ParentOp = CGN->getCallableRegion()->getParentOp();
+    llvm::dbgs() << ParentOp->getName() << " has "
+                 << DiscardableSymNodeUses[CGN] << " uses\n";
+  }
```
[the remark this answers is not in the document]
Comment (etiotto) [reply mlir-sycl/lib/Transforms/Inliner.cpp:489]: done

File: mlir-sycl/lib/Transforms/Inliner.cpp [hunk 6]
```diff
@@ -261,38 +547,59 @@ void Inliner::inlineCallsInSCC(Inliner &Inliner, CallGraphSCC &SCC) {
...
     if (failed(InlineRes)) {
       LLVM_DEBUG(llvm::dbgs() << "** Failed to inline\n");
       continue;
     }
-
     LLVM_DEBUG(llvm::dbgs() << "** Inline succeeded\n");
 
-    // then erase the call.
-    Call.erase();
+    NumInlinedCalls++;
```
Comment (whitneywhtsang) [inline mlir-sycl/lib/Transforms/Inliner.cpp:578]:
```suggestion
    ++NumInlinedCalls;
```

File: mlir-sycl/lib/Transforms/Inliner.cpp [hunk 7]
```diff
@@ -335,32 +644,102 @@ bool AlwaysInliner::shouldInline(ResolvedCall &ResolvedCall) const {
...
+  std::unique_ptr<InlinerBase> Inliner = nullptr;
+  switch (InlineMode) {
+  case sycl::InlineMode::AlwaysInline:
+    Inliner = std::make_unique<class AlwaysInliner>(Ctx, CG, SymTable);
+    break;
+  case sycl::InlineMode::Simple:
+    Inliner = std::make_unique<class Inliner>(Ctx, CG, SymTable);
+    break;
+  case sycl::InlineMode::Aggressive:
+    llvm_unreachable("TODO");
```
Comment (whitneywhtsang) [inline mlir-sycl/lib/Transforms/Inliner.cpp:711]: We may want to inline all SYCL device functions for aggressive. (No need to do in this PR.)

File: mlir-sycl/lib/Transforms/Inliner.cpp [hunk 8]
```diff
@@ -335,32 +644,102 @@ bool AlwaysInliner::shouldInline(ResolvedCall &ResolvedCall) const {
...
-  });
+  CGUseList UseList(getOperation(), CG, SymTable);
+
+  std::unique_ptr<InlinerBase> Inliner = nullptr;
+  switch (InlineMode) {
+  case sycl::InlineMode::AlwaysInline:
+    Inliner = std::make_unique<class AlwaysInliner>(Ctx, CG, SymTable);
+    break;
+  case sycl::InlineMode::Simple:
+    Inliner = std::make_unique<class Inliner>(Ctx, CG, SymTable);
```
Comment (whitneywhtsang) [inline mlir-sycl/lib/Transforms/Inliner.cpp:708]: optional:
```suggestion
    Inliner = std::make_unique<class SimpleInliner>(Ctx, CG, SymTable);
```

**Comment by etiotto** [comment]
> We are currently assuming this pass reaches a fixpoint after a single iteration. Could we need more than one? (can address on other PR).

I have addressed this comment in this PR.

File: polygeist/tools/cgeist/driver.cc (hunk 1 above)
Comment (etiotto) [reply polygeist/tools/cgeist/driver.cc]: done

File: mlir-sycl/lib/Transforms/Inliner.cpp (hunk 6 above)
Comment (etiotto) [reply mlir-sycl/lib/Transforms/Inliner.cpp:578]: In this context using pre-increment or post-increment doesn't make a difference.

File: mlir-sycl/lib/Transforms/Inliner.cpp [hunk 9]
```diff
@@ -309,16 +626,18 @@ void Inliner::collectCallOps(const CallGraphNode &SrcNode, const CallGraph &CG,
...
 //===----------------------------------------------------------------------===//
 
-/// Returns true if the given call should be inlined.
-bool AlwaysInliner::shouldInline(ResolvedCall &ResolvedCall) const {
-  // Don't allow inlining if the target is an ancestor of the call. This
-  // prevents inlining recursively.
-  if (ResolvedCall.TgtNode->getCallableRegion()->isAncestor(
-          ResolvedCall.Call->getParentRegion()))
+bool AlwaysInliner::shouldInline(ResolvedCall &ResolvedCall,
+                                 Optional<CGUseList> Uses) const {
```
Comment (whitneywhtsang) [inline mlir-sycl/lib/Transforms/Inliner.cpp:639]: No need to be optional, as it is always passed.
```suggestion
                                 CGUseList Uses) const {
```

File: mlir-sycl/lib/Transforms/Inliner.cpp (hunk 7 above)
Comment (etiotto) [reply mlir-sycl/lib/Transforms/Inliner.cpp:711]: I supposed that is one way to be 'aggressive', perhaps too aggressive. The aggressiveness we chose to implement will likely depends on whether we are inlining device code or host code. For the device we might be much more aggressive than on the host. We will need to experiment with the heuristic in a future PR.

File: mlir-sycl/lib/Transforms/Inliner.cpp [hunk 10]
```diff
@@ -226,29 +509,38 @@ runTransformOnSCCs(const CallGraph &CG,
...
-                                 MLIRContext *Ctx) {
-  Inliner::inlineCallsInSCC(Inliner, SCC);
+LogicalResult InlinerBase::inlineSCC(InlinerBase &Inliner, CGUseList &UseList,
+                                     CallGraphSCC &SCC,
+                                     Pass::Statistic &NumInlinedCalls) {
+  unsigned IterationCount = 0;
+  unsigned const constexpr MaxIterationCount = 5;
+  bool DidSomething = false;
+  do {
+    DidSomething |= inlineCallsInSCC(Inliner, UseList, SCC, NumInlinedCalls);
```
Comment (whitneywhtsang) [inline mlir-sycl/lib/Transforms/Inliner.cpp:519]:
```suggestion
    DidSomething = inlineCallsInSCC(Inliner, UseList, SCC, NumInlinedCalls);
```

File: mlir-sycl/lib/Transforms/Inliner.cpp (hunk 8 above)
Comment (etiotto) [reply mlir-sycl/lib/Transforms/Inliner.cpp:708]: Not for this PR, but I am starting to think that we could just have one Inliner class implementing the `shouldInline` member function based on heuristics. Or we could have an `Inliner` class that takes a `InlineHeuristic` object. That object would supply the heuristic to use.

Let's keep it like it is in this PR.

File: mlir-sycl/lib/Transforms/Inliner.cpp [hunk 11]
```diff
@@ -172,47 +283,219 @@ class Inliner : public InlinerInterface {
...
+  Operation *Op = CGN->getCallableRegion()->getParentOp();
+  if (!isa<SymbolOpInterface>(Op))
+    return isMemoryEffectFree(Op) && Op->use_empty();
+
+  // Otherwise, check the number of symbol uses.
+  auto SymbolIt = DiscardableSymNodeUses.find(CGN);
+  return SymbolIt != DiscardableSymNodeUses.end() && SymbolIt->second == 0;
+}
+
+bool CGUseList::hasOneUse(CallGraphNode *CGN) const {
```
Comment (whitneywhtsang) [inline mlir-sycl/lib/Transforms/Inliner.cpp:410]: I think the name in Inliner.cpp is better:
```suggestion
bool CGUseList::hasOneUseAndDiscardable(CallGraphNode *CGN) const {
```

File: mlir-sycl/lib/Transforms/Inliner.cpp [hunk 12]
```diff
@@ -512,11 +512,16 @@ LogicalResult InlinerBase::runTransformOnSCCs(
 LogicalResult InlinerBase::inlineSCC(InlinerBase &Inliner, CGUseList &UseList,
                                      CallGraphSCC &SCC,
                                      Pass::Statistic &NumInlinedCalls) {
-  inlineCallsInSCC(Inliner, UseList, SCC, NumInlinedCalls);
+  unsigned IterationCount = 0;
+  unsigned const constexpr MaxIterationCount = 5;
```
Comment (victor-eds) [inline mlir-sycl/lib/Transforms/Inliner.cpp:516]: For a future PR: Do we want to make this configurable?

**Review by victor-eds — APPROVED** [review]
After @whitneywhtsang comments, LGTM.

**Review by whitneywhtsang — APPROVED** [review]

**Pull request merged**